### PR TITLE
Implement tests for `block_accumulator`

### LINKED ISSUE #3346: Implement pruning of reactor's recent switch block headers

### PR DESCRIPTION
**Pull request by georgepisaltu**
Fixes #3346 

This PR adds tests for the `block_accumulator` component and its auxiliary parts.

### CODE REVIEW

File: node/src/components/block_accumulator.rs [hunk 1]
```diff
@@ -808,6 +811,9 @@ impl<REv: ReactorEvent> Component<REv> for BlockAccumulator {
             Event::ExecutedBlock { hot_block } => {
                 let height = hot_block.block.header().height();
                 let era_id = hot_block.block.header().era_id();
+                // Logically it makes sense to purge the acceptors here or in
+                // `register_local_tip`, but only on the path where a new tip
+                // is accepted.
                 self.register_local_tip(height, era_id);
```
Comment (EdHastingsCasperAssociation) [inline node/src/components/block_accumulator.rs]: we do purge here (first line inside `register_local_tip`)

I think you are correct that the purge could be moved to within new block detection (`if new_local_tip`) tho

File: node/src/components/block_accumulator/block_acceptor.rs [hunk 2]
```diff
@@ -389,3 +389,39 @@ impl BlockAcceptor {
...
+    pub(super) fn signatures(&self) -> &BTreeMap<PublicKey, (FinalitySignature, BTreeSet<NodeId>)> {
+        &self.signatures
+    }
+
+    #[allow(unused)]
+    pub(super) fn signatures_mut(
+        &mut self,
+    ) -> &mut BTreeMap<PublicKey, (FinalitySignature, BTreeSet<NodeId>)> {
+        &mut self.signatures
+    }
```
Comment (EdHastingsCasperAssociation) [inline node/src/components/block_accumulator/block_acceptor.rs]: did you mean to leave #[allow(unused)] on this?

File: node/src/components/block_accumulator.rs (hunk 1 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator.rs]: Yes, that is what I meant there.

File: node/src/components/block_accumulator/block_acceptor.rs (hunk 2 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/block_acceptor.rs]: I did not, it's a remnant of a previous iteration and I'll remove it.

**Review by EdHastingsCasperAssociation — APPROVED** [review]

File: node/src/components/block_accumulator/tests.rs [hunk 3]
```diff
@@ -1,9 +1,272 @@
...
+            Event::NetworkPeerBehaviorAnnouncement(peer_behavior) => {
+                self.blocked_peers.push(peer_behavior);
+                Effects::new()
+            }
+        }
+    }
+}
 
 #[test]
 fn upsert_acceptor() {
```
Comment (rafal-ch) [inline node/src/components/block_accumulator/tests.rs:269]: I know that this test is not added in this PR, but while you're here, can you consider extending this test so that it also covers the block timestamp purging?

I see that when `block_timestamps.pop_front()` line is commented, all tests still pass.

File: node/src/components/block_accumulator/tests.rs [hunk 4]
```diff
@@ -61,3 +324,768 @@ fn upsert_acceptor() {
...
+    // should not store anything.
+    acceptor.set_sufficient_finality(true);
+    let (should_store, _offenders) = acceptor.should_store_block(&era_validator_weights);
+    assert_eq!(should_store, ShouldStore::Nothing);
+    // Reset the flag.
+    acceptor.set_sufficient_finality(false);
+
+    let (should_store, offenders) = acceptor.should_store_block(&era_validator_weights);
+    assert_eq!(should_store, ShouldStore::Nothing);
+    assert!(offenders.is_empty());
```
Comment (rafal-ch) [inline node/src/components/block_accumulator/tests.rs:637]: Can we also have a test case when offenders are not expected to be empty?

File: node/src/components/block_accumulator/tests.rs [hunk 5]
```diff
@@ -61,3 +324,768 @@ fn upsert_acceptor() {
...
+                let event = super::Event::ExecutedBlock {
+                    hot_block: hot_block_1.clone(),
+                };
+                effect_builder
+                    .into_inner()
+                    .schedule(event, QueueKind::Validation)
+                    .ignore()
+            })
+            .await;
+        for _ in 0..4 {
```
Comment (rafal-ch) [inline node/src/components/block_accumulator/tests.rs:1400]: Is there a specific number of times we need to crank? I see that it's set to `4` here, but `6` in other cases.
If yes, please consider using a named constant.

File: node/src/components/block_accumulator.rs [hunk 6]
```diff
@@ -328,6 +328,9 @@ impl BlockAccumulator {
             .as_ref()
             .map_or(false, |local_tip| block_height < local_tip.height)
         {
+            // If the acceptor exists already (through a previously recorded
+            // signature), we should probably purge it if it's older than
```
Comment (rafal-ch) [inline node/src/components/block_accumulator.rs]: I think we should clarify this "probably" before merging this PR, wdyt?

File: node/src/components/block_accumulator/tests.rs [hunk 7]
```diff
@@ -61,3 +324,768 @@ fn upsert_acceptor() {
...
+    assert_eq!(should_store, ShouldStore::Nothing);
+
+    // Create the third validator's signature.
+    let fin_sig = FinalitySignature::create(
+        *block.hash(),
+        block.header().era_id(),
+        &keys[2].0,
+        keys[2].1.clone(),
+    );
+    // The third signature with weight 10% doesn't change make the block go to
```
Comment (Fraser999) [inline node/src/components/block_accumulator/tests.rs]:
```suggestion
    // The third signature with weight 10% doesn't make the block go to
```

File: node/src/components/block_accumulator.rs [hunk 8]
```diff
@@ -328,6 +328,9 @@ impl BlockAccumulator {
...
+            // the local tip.
```
Comment (Fraser999) [inline node/src/components/block_accumulator.rs]: Won't the next purge do that anyway?

File: node/src/components/block_accumulator/tests.rs [hunk 9]
```diff
@@ -1,9 +1,272 @@
-use casper_types::testing::TestRng;
+use std::{
```
Comment (rafal-ch) [inline node/src/components/block_accumulator/tests.rs:1]: I think it might be beneficial to cover some utility functions with simple UTs, for example, I can flip the `>` comparison in `should_leap()` and all tests still pass. Such change will probably be detected in NCTL tests, but these are not that quick to run.

Also, `purge()` has a potential to contain some off-by-one errors.

File: node/src/components/block_accumulator/tests.rs [hunk 10]
```diff
@@ -61,3 +324,768 @@ fn upsert_acceptor() {
...
+        };
+        let effects = block_accumulator.handle_event(effect_builder, &mut rng, event);
+        assert!(effects.is_empty());
+        // This should *probably* have no effect on the accumulator since the
+        // block is older than the local tip, but right now it creates the
+        // acceptor if it's in the same era or newer than the local tip era.
+        // See comment in `BlockAccumulator::register_block`.
+        // assert!(!block_accumulator
+        //     .block_acceptors
+        //     .contains_key(older_block.hash()));
```
Comment (Fraser999) [inline node/src/components/block_accumulator/tests.rs]: This is registering a finality signature though rather than a block.

You could maybe provide a signature from an older era and check no new acceptor is created, and/or uncomment this but assert that the acceptor exists (with a comment about why it does).

**Review by Fraser999 — COMMENTED** [review]
Generally looking good, but just a couple of comments about the comments :)

File: node/src/components/block_accumulator.rs (hunk 6 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator.rs]: I think we should, yes. I put this comment here to get input from multiple people.

File: node/src/components/block_accumulator.rs (hunk 8 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator.rs]: It will, but at the same time why not purge it at the exact time you find out it's no longer useful?

File: node/src/components/block_accumulator/tests.rs (hunk 9 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:1]: The operator flip works like that exactly because `should_leap` doesn't have unit tests. I agree we should have those, but at the same time there are lots of functions that don't have unit tests yet (it's a lot of code). In this PR I wanted to test the most *problematic* things in the accumulator. We can incrementally add more and more tests, and I will add the `should_leap` unit test along with the `purge` in the next iteration of this PR.

File: node/src/components/block_accumulator/tests.rs (hunk 3 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:269]: The explanation for this is the same as the other comment. I will extend the test.

File: node/src/components/block_accumulator/tests.rs (hunk 4 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:637]: Will do.

File: node/src/components/block_accumulator/tests.rs (hunk 5 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:1400]: These numbers are not constant, they actually represent the number of events that will be processed by scheduling the above event and they are not constant, not even for the same scheduled event. Unfortunately I don't see a nice way to handle this. I can add comments above the `try_cranks` explaining that this is how many events will be handled, but not much more.

File: node/src/components/block_accumulator/tests.rs (hunk 10 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs]: The problem scenario is this:
- we have an old block which, if received by the accumulator, would not have an acceptor created
- if you start by sending the signatures for said block, which creates the acceptor if the era is recent enough
- only after receiving the block we realize that this acceptor is useless
- it will eventually get purged

This is wasteful and arguably introduces unnecessary complexity. We could keep track of purged block hashes until their era stops being recent, but this would probably not be worth the hassle. Alternatively, we could just purge when we realize it's a stale block and live with it.

File: node/src/components/block_accumulator.rs (hunk 8 above)
Comment (Fraser999) [reply node/src/components/block_accumulator.rs]: I'd lean towards keeping the code simpler.  Also, the purge method looks like it could be expensive and if we're only calling from one place it makes it easier to rationalize about.  So I'd favour just deleting this comment and leaving purge as is.

File: node/src/components/block_accumulator/tests.rs (hunk 9 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:1]: Done.

File: node/src/components/block_accumulator/tests.rs (hunk 3 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:269]: Done.

File: node/src/components/block_accumulator/tests.rs (hunk 4 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs:637]: Done.

File: node/src/components/block_accumulator.rs (hunk 6 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator.rs]: Talked with @Fraser999 and we'll keep it as is.

File: node/src/components/block_accumulator/tests.rs (hunk 7 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs]: Done.

File: node/src/components/block_accumulator/tests.rs (hunk 10 above)
Comment (georgepisaltu) [reply node/src/components/block_accumulator/tests.rs]: We decided not to do anything about this. We accept this behavior to be the expected one.

**Review by Fraser999 — APPROVED** [review]

**Review by rafal-ch — APPROVED** [review]
LGTM :+1:
